### ISSUE TITLE
Lamps not identified as lamps with f/w 1.3.14

### ISSUE THREAD

**Issue by mhlund** [ggravlingen/pytradfri#135]
I am a new user of pytradfri trying to integrate my gateway with homeassistant.
As I know HA only use version 4.1.0 of pytradfri I have also been testing with the standalone version of pytradfri 5.2.0

As my gateway is new and autoupgraded to v 1.3.14 imediately I have'nt tested any earlier firmwares of the gateway. My lamps have fw-version 1.2.217.

When I run pytradfri in Homeassistant my lamps are detected as sensors. I have the same problem with my standalone 5.2.0 test.

output from "python3 -i -m pytradfri 192.168.1.7 key"

```
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001']
DEBUG:pytradfri.api.libcoap_api:Received: [65536,65537,65539,65538,65540]
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', 65536]
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Fjärrkontroll 1","5750":0,"9002":1515435028,"9020":1515870426,"9003":65536,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI remote control","2":"","3":"1.2.214","6":3,"9":87},"15009":[{"9003":0}]}
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', 65537]
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Tak vr 1","5750":0,"9002":1515435088,"9020":1516008067,"9003":65537,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb GU10 WS 400lm","2":"","3":"1.2.217","6":1},"15009":[{"5851":10,"5850":0,"5717":0,"5711":450,"5709":30138,"5710":26909,"5706":"efd275","9003":0}]}
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', 65539]
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Tak vr 3","5750":0,"9002":1515435277,"9020":1516008127,"9003":65539,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb GU10 WS 400lm","2":"","3":"1.2.217","6":1},"15009":[{"5851":10,"5850":0,"5717":0,"5711":450,"5709":30138,"5710":26909,"5706":"efd275","9003":0}]}
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', 65538]
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Tak vr 2","5750":0,"9002":1515435198,"9020":1516008063,"9003":65538,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb GU10 WS 400lm","2":"","3":"1.2.217","6":1},"15009":[{"5851":10,"5850":0,"5717":0,"5711":450,"5709":30138,"5710":26909,"5706":"efd275","9003":0}]}
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', 65540]
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Tak vr 4","5750":0,"9002":1515435391,"9020":1516008285,"9003":65540,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb GU10 WS 400lm","2":"","3":"1.2.217","6":1},"15009":[{"5851":10,"5850":0,"5717":0,"5711":450,"5709":30138,"5710":26909,"5706":"efd275","9003":0}]}
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.6/dist-packages/pytradfri/__main__.py", line 36, in <module>
    light = lights[0]
IndexError: list index out of range
>>> lights
[]
>>>
```

I would expect my 4 TRADFRI bulb GU10 WS 400lm to be identified as lamps, but the lights array looks empty. I've tried to read the source, but is new to python (c# and perl nerd).

    gateway = Gateway()
    devices_commands = api(gateway.get_devices())
    devices = api(devices_commands)
    lights = [dev for dev in devices if dev.has_light_control]
    light = lights[0]
 
It looks to me as pytradfri when calling the coap api does'nt get it right. The lamp devices does not have "has_light_control" set to true...

**Comment by ggravlingen** [comment]
One quick note, 5.2.0 is not compatible with Home Assistant, as its integration is built around logic that was present in 4.1.0 but has since then been depreciated.

It doesn't appear that your lights return 3311 in the json, which is what's required to get has_light_control set to true. Is any of your bulbs lit up, or has ever been lit up?

Pretty formatted json for further analysis:
```
{  
   "9001":"Tak vr 4",
   "5750":0,
   "9002":1515435391,
   "9020":1516008285,
   "9003":65540,
   "9054":0,
   "9019":1,
   "3":{  
      "0":"IKEA of Sweden",
      "1":"TRADFRI bulb GU10 WS 400lm",
      "2":"",
      "3":"1.2.217",
      "6":1
   },
   "15009":[  
      {  
         "5851":10,
         "5850":0,
         "5717":0,
         "5711":450,
         "5709":30138,
         "5710":26909,
         "5706":"efd275",
         "9003":0
      }
   ]
}```

**Comment by ggravlingen** [comment]
@lwis collecting your thoughts here as well

 I haven't seen 15009 in a bulb before but then again, I don't think any of us have used this type of (bulb)[http://www.ikea.com/se/sv/images/products/tradfri-led-ljuskalla-gu-lumen__0463170_PE608601_S4.JPG] before either? I've been under the impression that 15009 references power switches but that not seems to be an incorrect assumption. Looks like this a new way of representing a bulb by the gateway?

**Comment by Lakitna** [comment]
Do your bulbs show up if you run print `devices` instead of `lights`?

That will print all devices on the network, including remotes and such. I'm assuming they will show up here, but some confirmation never hurts.

**Comment by mhlund** [comment]
Yes, they show up if I print `devices` instead of `lights`!

`>>> lights
[]
>>> devices
[<65536 - Fjärrkontroll 1 (TRADFRI remote control)>, <65537 - Tak vr 1 (TRADFRI bulb GU10 WS 400lm)>, <65539 - Tak vr 3 (TRADFRI bulb GU10 WS 400lm)>, <65538 - Tak vr 2 (TRADFRI bulb GU10 WS 400lm)>, <65540 - Tak vr 4 (TRADFRI bulb GU10 WS 400lm)>]`

I'm aware that Homeassistant uses v 4.1.0 of pytradfri (which by the way has the same behaviour)

**Comment by mhlund** [comment]
If i commented out row 36 "light = lights[0]" in __main__.py I was able to run dump_all().
Here's the result!

```
>>> dump_all()
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['.well-known', 'core']
DEBUG:pytradfri.api.libcoap_api:Received: <//15006>;ct=0;obs,<//15001>;ct=0;obs,<//15004>;ct=0;obs,<//15004/add>;ct=0,<//15004/remove>;ct=0,<//15010>;ct=0;obs,<//15005>;ct=0;obs,<//15011/15012>;ct=0;obs,<//15011/9034>;ct=0,<//15011/9030>;ct=0,<//15011/9031>;ct=0,<//15011/9094>;ct=0;obs,<//15011/9095>;ct=0;obs,<//15011/9104>;ct=0;obs,<//15004/131073>;ct=0;obs,<//15005/131073>;ct=0;obs,<//15001/65536>;ct=0;obs,<//15001/65537>;ct=0;obs,<//15001/65539>;ct=0;obs,<//15001/65538>;ct=0;obs,<//15001/65540>;ct=0;obs,<//15005/131073/196608>;ct=0;obs,<//15005/131073/196609>;ct=0;obs,<//15005/131073/196610>;ct=0;obs,<//15011/9063>;ct=0
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15006']
DEBUG:pytradfri.api.libcoap_api:Received: [{"9002":1515921628,"9014":0,"9015":5001}]
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001']
DEBUG:pytradfri.api.libcoap_api:Received: [65536,65537,65539,65538,65540]
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15004']
DEBUG:pytradfri.api.libcoap_api:Received: [131073]
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15010']
DEBUG:pytradfri.api.libcoap_api:Received: []
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15005']
DEBUG:pytradfri.api.libcoap_api:Received: [131073]
None

DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15011', '15012']
DEBUG:pytradfri.api.libcoap_api:Received: {"9023":"2.tradfri.pool.ntp.org","9059":1516024009,"9073":0,"9060":"2018-01-15T13:46:49.005464Z","9080":0,"9071":1,"9062":0,"9061":0,"9093":0,"9029":"1.3.14","9081":"7e1a55520440004d","9092":0,"9069":1515425361,"9082":true,"9055":0,"9083":"061-05-989","9066":5,"9054":0,"9077":0,"9072":0,"9074":0,"9075":0,"9076":0,"9078":0,"9079":0,"9106":0,"9105":0}
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15011', '9034']
DEBUG:pytradfri.api.libcoap_api:Received:
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15011', '9030']
DEBUG:pytradfri.api.libcoap_api:Received:
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15011', '9031']
DEBUG:pytradfri.api.libcoap_api:Received:
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15011', '9094']
DEBUG:pytradfri.api.libcoap_api:Received:
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15011', '9095']
DEBUG:pytradfri.api.libcoap_api:Received:
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15011', '9104']
DEBUG:pytradfri.api.libcoap_api:Received:
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15004', '131073']
DEBUG:pytradfri.api.libcoap_api:Received: {"9018":{"15002":{"9003":[65536,65537,65538,65539,65540]}},"9001":"Taklampa vardagsrum","9002":1515435031,"9039":196608,"9003":131073,"5851":10,"5850":1}
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15005', '131073']
DEBUG:pytradfri.api.libcoap_api:Received: [196608,196609,196610]
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', '65536']
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Fjärrkontroll 1","5750":0,"9002":1515435028,"9020":1515870426,"9003":65536,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI remote control","2":"","3":"1.2.214","6":3,"9":87},"15009":[{"9003":0}]}
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', '65537']
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Tak vr 1","5750":0,"9002":1515435088,"9020":1516008067,"9003":65537,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb GU10 WS 400lm","2":"","3":"1.2.217","6":1},"15009":[{"5851":10,"5850":1,"5717":0,"5711":450,"5709":30138,"5710":26909,"5706":"efd275","9003":0}]}
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', '65539']
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Tak vr 3","5750":0,"9002":1515435277,"9020":1516008127,"9003":65539,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb GU10 WS 400lm","2":"","3":"1.2.217","6":1},"15009":[{"5851":10,"5850":1,"5717":0,"5711":450,"5709":30138,"5710":26909,"5706":"efd275","9003":0}]}
None

DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', '65538']
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Tak vr 2","5750":0,"9002":1515435198,"9020":1516008063,"9003":65538,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb GU10 WS 400lm","2":"","3":"1.2.217","6":1},"15009":[{"5851":10,"5850":1,"5717":0,"5711":450,"5709":30138,"5710":26909,"5706":"efd275","9003":0}]}
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', '65540']
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Tak vr 4","5750":0,"9002":1515435391,"9020":1516008285,"9003":65540,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb GU10 WS 400lm","2":"","3":"1.2.217","6":1},"15009":[{"5851":10,"5850":1,"5717":0,"5711":450,"5709":30138,"5710":26909,"5706":"efd275","9003":0}]}
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15005', '131073', '196608']
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"EVERYDAY","9002":1515435031,"9003":196608,"9057":0,"9068":1}
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15005', '131073', '196609']
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"RELAX","9002":1515435031,"9003":196609,"9057":1,"9068":1}
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15005', '131073', '196610']
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"FOCUS","9002":1515435031,"9003":196610,"9057":2,"9068":1}
None


DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15011', '9063']
DEBUG:pytradfri.api.libcoap_api:Received:
None

```

**Comment by mhlund** [comment]
Is there anything more I can do to help solve this problem?

**Comment by Lakitna** [comment]
At the moment I don't think so. Once there is a PR you can help by testing it, but right now @ggravlingen or @lwis have to look into it and find a new, better way to separate lamps from controllers.

They should be able to do that with the information you've provided earlier.

**Comment by lwis** [comment]
So, it doesn't appear the GU10 bulbs return 3311? Very strange, indeed.

@ggravlingen do you have a 3311 handy to compare with the 15009 payload?

**Comment by ggravlingen** [comment]
@lwis examples are available here: https://github.com/ggravlingen/pytradfri/blob/ee5a45fbe099463418f5f64bd8f6fdc7a2c8e6d1/tests/devices.py#L8

```
LIGHT_W = {
  '3': {'0': 'IKEA of Sweden',
        '1': 'TRADFRI bulb E27 W opal 1000lm',
        '2': '',
        '3': '1.2.214',
        '6': 1},
  '3311': [{'5850': 1, '5851': 254, '9003': 0}],
  '5750': 2,
  '9001': 'Hall 1',
  '9002': 1509923551,
  '9003': 65537,
  '9019': 1,
  '9020': 1510009959,
  '9054': 0
}
```

**Comment by mhlund** [comment]
The only pattern I can identify is that the array "3"'s member "6" is 1 for lamps and 3 for remotes as far as the data I have... But I do not have all kinds of bulbs, remotes and sensors, so I could be completely wrong!
Scenes and Groups do not have the "3"-array at all...

**Comment by ggravlingen** [comment]
@mhlund if you look at the json of a regular white buld you'll see this, where each of the numbers (eg 5850, 5851) represents some information. Check const.py for more info on what they mean.
```
   '3311': [{
             '5706': '0',
             '5709': 31103,
             '5710': 27007,
             '5711': 400,
             '5850': 1,
             '5851': 254,
             '9003': 0}],
```

Your light returns (although under 15009 and not 3311) this information:
```
   "15009":[  
      {  
         "5851":10,
         "5850":0,
         "5717":0,
         "5711":450,
         "5709":30138,
         "5710":26909,
         "5706":"efd275",
         "9003":0
      }
```

As you see, it's quite similar information so the fix mig not be that difficult to achieve.

**Comment by ggravlingen** [comment]
I'm discussing possible solutions with @lwis

**Comment by ggravlingen** [comment]
Started looking into a solution #141

**Comment by mhlund** [comment]
Finns det nåt enkelt sätt att testa de förändringar som gjorts nu, eller är det enklast att vänta på att pull requesten går igenom och det landar i master branch'en? :)

**Comment by ggravlingen** [comment]
Write in English here, please. Even though I understand what you write the other two collaborators of pytradfri do not. 😊

To answer your question (if my PR can be tested): yes you can fetch my branch or simply replace the device.py file with my version and give it a go. It should work and I would appreciate if you can test it since I don't have gu10 lights.

**Comment by ggravlingen** [comment]
Download this file and replace device.py with it and then run command line interface: https://github.com/ggravlingen/pytradfri/blob/3e89568b545da487175530f9336c1728e53dd67e/pytradfri/device.py

**Comment by mhlund** [comment]
Ahh, I did'nt even think about language as I wrote... :) 
I must be a bit to relaxed here in the sofa watching Star Trek this sunday afternoon!

I'll try downloading the file for a test drive!

**Comment by mhlund** [comment]
I now get the following error when testing...

```
mho@srv1:~$ python3 -i -m pytradfri 192.168.1.7
DEBUG:pytradfri.util:JSON file not found: tradfri_standalone_psk.conf
Please provide the 'Security Code' on the back of your Tradfri gateway: XXXXXXXXXXXXXX
DEBUG:pytradfri.util:JSON file not found: tradfri_standalone_psk.conf
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 post ['15011', '9063']: {'9090': '02b35a0dd3e84a5fb04c81138bf5f51b'}
DEBUG:pytradfri.api.libcoap_api:Received: {"9091":"XXXXXXXXXXXXXX","9029":"1.3.0014"}
Generated PSK:  XXXXXXXXXXXXXX
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001']
DEBUG:pytradfri.api.libcoap_api:Received: [65536,65537,65539,65538,65540]
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', 65536]
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Fjärrkontroll 1","5750":0,"9002":1515435028,"9020":1516476093,"9003":65536,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI remote control","2":"","3":"1.2.214","6":3,"9":87},"15009":[{"9003":0}]}
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', 65537]
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Tak vr 1","5750":0,"9002":1515435088,"9020":1516526475,"9003":65537,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb GU10 WS 400lm","2":"","3":"1.2.217","6":1},"15009":[{"5851":5,"5850":0,"5717":0,"5711":0,"5709":32886,"5710":27217,"5706":"efd275","9003":0}]}
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', 65539]
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Tak vr 3","5750":0,"9002":1515435277,"9020":1516526530,"9003":65539,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb GU10 WS 400lm","2":"","3":"1.2.217","6":1},"15009":[{"5851":5,"5850":0,"5717":0,"5711":0,"5709":32886,"5710":27217,"5706":"efd275","9003":0}]}
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', 65538]
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Tak vr 2","5750":0,"9002":1515435198,"9020":1516526482,"9003":65538,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb GU10 WS 400lm","2":"","3":"1.2.217","6":1},"15009":[{"5851":5,"5850":0,"5717":0,"5711":0,"5709":32886,"5710":27217,"5706":"efd275","9003":0}]}
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001', 65540]
DEBUG:pytradfri.api.libcoap_api:Received: {"9001":"Tak vr 4","5750":0,"9002":1515435391,"9020":1516531869,"9003":65540,"9054":0,"9019":1,"3":{"0":"IKEA of Sweden","1":"TRADFRI bulb GU10 WS 400lm","2":"","3":"1.2.217","6":1},"15009":[{"5851":5,"5850":0,"5717":0,"5711":0,"5709":32886,"5710":27217,"5706":"efd275","9003":0}]}
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15004']
DEBUG:pytradfri.api.libcoap_api:Received: [131073]
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15005', <pytradfri.command.Command object at 0x7f22b2c01668>]
DEBUG:pytradfri.api.libcoap_api:Received: [131073]
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15010']
DEBUG:pytradfri.api.libcoap_api:Received: []
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15011', '15012']
DEBUG:pytradfri.api.libcoap_api:Received: {"9023":"2.tradfri.pool.ntp.org","9059":1516546898,"9073":0,"9060":"2018-01-21T15:01:38.004496Z","9080":0,"9071":1,"9062":0,"9061":0,"9093":0,"9029":"1.3.14","9081":"7e1a55520440004d","9092":0,"9069":1515425361,"9082":true,"9055":0,"9083":"061-05-989","9066":5,"9054":0,"9077":0,"9072":0,"9074":0,"9075":0,"9076":0,"9078":0,"9079":0,"9106":0,"9105":0}

Example commands:
> devices
> homekit_id
> light.light_control.lights
> api(light.light_control.set_dimmer(10))
> api(light.light_control.set_dimmer(254))
> api(light.light_control.set_xy_color(254))
> api(lights[1].light_control.set_dimmer(20))
> tasks[0].repeat_days_list
> api(gateway.reboot())
> groups
> moods
> tasks
> dump_devices()
> dump_all()
>>> lights
[<65536 - Fjärrkontroll 1 (TRADFRI remote control)>, <65537 - Tak vr 1 (TRADFRI bulb GU10 WS 400lm)>, <65539 - Tak vr 3 (TRADFRI bulb GU10 WS 400lm)>, <65538 - Tak vr 2 (TRADFRI bulb GU10 WS 400lm)>, <65540 - Tak vr 4 (TRADFRI bulb GU10 WS 400lm)>]
>>>
```
As far as I'm able to decode this it looks ok! :+1:

**Comment by ggravlingen** [comment]
Do you mean the file not found error?

**Comment by ggravlingen** [comment]
I didn't get that error, but then again I had a tradfri_standalone_psk.conf file in my pytradfri folder. I assume you didn't?

**Comment by mhlund** [comment]
Oops... Sorry... I started writing after only replacing device.py and got an import error... I got that because I had version 5.2.0... Then after upgrading to v5.2.2 everything worked, but the first sentence was left in my text... It should have read "I now get the following RESULT when testing" :)

Everything looks ok for me now!

Sorry about the confusion!

**Comment by ferdydek** [comment]
I'm probably super late with the feedback ...
I've been using GU10 (tradfri_bulb_gu10_ws_400lm - the type with changeable temperature of the light) lamps for more than half a year and had no issues with them (via home-assistant). My GW is 1.3.14, bulbs 1.2.217.

**Comment by Lakitna** [comment]
@ferdydek That's weird, can you provide the raw output of one of these lamps? You can use code similar to this to achieve that:

```
lights = [dev for dev in devices if dev.has_light_control]
lights[0].raw  # Change 0 to index of GU10 light
```

**Comment by ferdydek** [comment]
yeah sure just need to find someplace to install it. will update soon.

**Comment by ggravlingen** [comment]
This is odd, looks like IKEA handles GU10 bulbs in two different ways in the gateway.

**Comment by lwis** [comment]
ferdydek's 5750 value is '2' - that seems to be the only obvious difference I'm seeing between the two payloads? Do we know what 'application_type' is?

**Comment by ferdydek** [comment]
there are 2 GU10 types, I only have the first one:
http://www.ikea.com/pt/en/catalog/products/00318292/     -- model: **LED1650R5** 
http://www.ikea.com/pt/en/catalog/products/00318292/     -- model: **LED1537R6** 
maybe that is the difference ? As one of them supports color temperature change and the other has fixed color temperature?

**Comment by lwis** [comment]
Both are 'TRADFRI bulb GU10 WS 400lm' indicating that they are both white spectrum.

**Comment by Lakitna** [comment]
You'd expect that this distinction would follow the same pattern as with E27 WS and W bulbs.

White bulb E27 (no color):

```
{
  "3": {
    "0": "IKEA of Sweden",
    "1": "TRADFRI bulb E27 opal 1000lm",
    "2": "",
    "3": "1.2.214",
    "6": 1
  },
  "3311": [
    {
      "5850": 0,
      "5851": 254,
      "9003": 0
    }
  ],
  "5750": 2,
  "9001": "Hallway",
  "9002": 1508581972,
  "9003": 65542,
  "9019": 1,
  "9020": 1516550605,
  "9054": 0
}
```

White color spectrum E27:

```
{
  "3": {
    "0": "IKEA of Sweden",
    "1": "TRADFRI bulb E27 WS opal 980lm",
    "2": "",
    "3": "1.2.217",
    "6": 1
  },
  "3311": [
    {
      "5706": "f5faf6",
      "5709": 24841,
      "5710": 24593,
      "5711": 250,
      "5717": 0,
      "5850": 1,
      "5851": 254,
      "9003": 0
    }
  ],
  "5750": 2,
  "9001": "Sander plafond",
  "9002": 1504787946,
  "9003": 65537,
  "9019": 1,
  "9020": 1516537561,
  "9054": 0
}
```

No weird differences with these bulbs.

Could this be a regional thing? Or maybe different production runs?

**Comment by ferdydek** [comment]
The one from first link is white **spectrum**, the second is just while **fixed**.
And the second one also is a "new" product, was not part of initial tradfri products group.

**Comment by Lakitna** [comment]
I didn't know about that one yet, but it seems like both you and @mhlund are using GU10 White spectrum bulbs (note that in both rawOutput[3][1] have WS in them)

**Comment by Lakitna** [comment]
You're both using the same type of lamps (key `1`) according to the dumps. You are also using the same firmware (key `3`) and both your lamps are produced by Ikea (key `0`).

The bottom line is that there is no apparent reason why @mhlund his lamp uses `15009` where @ferdydek his lamp uses `3311`

@mhlund 
```
  '3': {
    '0': 'IKEA of Sweden', 
    '1': 'TRADFRI bulb GU10 WS 400lm', 
    '2': '', 
    '3': '1.2.217', 
    '6': 1
  },
```

@ferdydek 
```
   '3':{  
      '0':'IKEA of Sweden',
      '1':'TRADFRI bulb GU10 WS 400lm',
      '2':'',
      '3':'1.2.217',
      '6':1
   }
```

**Comment by Lakitna** [comment]
@lwis noted that the value of the `5750` is different. I've done a little digging and here is what I found:

From const.py:
```
ATTR_APPLICATION_TYPE = "5750"
```

So it should denote something like the purpose of the device. After printing the raw output of all my devices (not just lights) I found that `5750` can have a few different values:

0 = Remote control
1 = ??? I think dimmer remote, but I don't have any so I can't check
2 = Light bulb
3 = ??? Or maybe this is the dimmer remote? Could panels be a separate category?
4 = Motion sensor

With @mhlund his lamp `5750` is 0, suggesting that it's a remote control in the eyes of the Gateway. Like all other devices with `5750` = 0 it uses `15009` to store it's values. Light bulbs (`5750` = 2) do this in `3311`.

In conclusion. Somehow the Gateway of @mhlund thinks that his lamps are remote controls. I suggest repairing, that might repair the issue.

**Comment by mhlund** [comment]
That's really strange! Could this have happened because @ferdydek paired his lamps with his gateway with a previous f/w version? My gateway is new and it was upgraded to f/w 1.3.14 before I paired my GU10's

**Comment by Lakitna** [comment]
Possibly, it's also possible something went wrong during your pairing process, possibly because of a bug in the Gateway firmware or due to the old firmware on your bulbs (they can only be updated after you pair them).

I'd like to exclude as many things as possible right now, so could you please try re-pairing one of your lamps and post the raw output?

**Comment by Lakitna** [comment]
> How do I perform a "repair"?

English is a funny language sometimes. I meant redo the pairing process, or re-pair. You can simply do this by deleting one of the lamps using the app and pair it once more via the same app.

**Comment by ferdydek** [comment]
Worth mentioning, after the last year's gateway auto-update I actually ended up doing clean setup from scratch (remove all devices, factory reset, add all devices).
It's not the original FW, @mhlund, as my bulbs were purchased on 3 diff times and actually in portugal and in spain, so some were with 214 some with 217 FW. I don't think the bulbs have anything to do here, maybe GW or pairing process - I used the normal pairing with remote, not the trick where ppl connect bulbs directly to GW.

**Comment by Lakitna** [comment]
Do you happen to remember what Gateway version this was? I'm guessing 1.2.x, the update where security was changed.

**Comment by ferdydek** [comment]
Yes the one that broke compatibility with pytradfri. 1.2.42.

**Comment by Lakitna** [comment]
About two weeks ago Ikea pushed 1.3.14. That could be the problem here.

I'll pair a new E27 lamp to find out if this is simply a new way of handling values on Ikea's part.

**Comment by Lakitna** [comment]
Pairing a new E27 WS bulb (FW 1.2.217) results in a normal situation where `5750` = 2 and `3311` is used to store values.

**Comment by schliflo** [comment]
I have a bunch of GU10 lights here (Germany). Let me know if I can help debugging ;)

**Comment by mhlund** [comment]
I did a factory reset of my gateway and now I am not able do the initial pairing of my remote...
After I reset my remote with 4 "clicks" and waiting 10 seconds before holding the pairing button on the remote until the gateway blinks 4 times my iphone app just keeps searching for the unit...

I'm not able to connect pytradfri againt the gateway either... 

```
mho@srv1:~$ python3 -i -m pytradfri 192.168.1.7
DEBUG:pytradfri.api.libcoap_api:Executing 192.168.1.7 get ['15001']
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.6/dist-packages/pytradfri/__main__.py", line 61, in <module>
    devices_commands = api(gateway.get_devices())
  File "/usr/local/lib/python3.6/dist-packages/pytradfri/api/libcoap_api.py", line 92, in request
    return self._execute(api_commands)
  File "/usr/local/lib/python3.6/dist-packages/pytradfri/api/libcoap_api.py", line 81, in _execute
    raise RequestTimeout() from None
pytradfri.error.RequestTimeout
>>>
```
Strange!

**Comment by ggravlingen** [comment]
@mhlund are you not back on an earlier version of gw after reset?

**Comment by mhlund** [comment]
I can't tell as I can't communicate with the gateway... In the ios app the only thing I can do is an initial pairing with a remote, and that does'nt work now...

**Comment by ferdydek** [comment]
What worked for me few months back was to pair phone with GW, then in the APP "clear all settings" then "disconnect from gw", then factory reset and re-pairing or everything from scratch.

**Comment by mhlund** [comment]
Exactly what I did @ferdydek ! First a "clear all settings" in the app, then a factory reset on the gateway and then reconnecting the app by scanning the QR code and then trying to pair my remote.... The last step unsuccessful...

**Comment by ferdydek** [comment]
Did you factory reset the remote as well? I think its 4 presses in 5 seconds.

**Comment by mhlund** [comment]
Yes! And then when I hold the remote pairing button in for 5 seconds the gateway blinks 4 times as expected, but the app still searches for it... It never shows up!

**Comment by ferdydek** [comment]
Any luck with getting it back up @mhlund ? Or its going back to Ikea ?

**Comment by mhlund** [comment]
I've just completed my setup, after more than 10 factory resets and then waaaaaiting for a loooong time I was able to pair my remote with the gateway! ... and when that was done I could pair my lamps!

And lo and behold! Suddenly my lamps return the 3311 structure as supposed to! Very strange!

My gateway firmware still is 1.3.14 and lamps are 1.2.217... Very strange bahviour indeed!

```
>>> lights[1].raw
{
  '9001': 'Taklampa VR1',
  '9002': 1516659142,
  '9020': 1516659151,
  '9003': 65539,
  '9054': 0,
  '5750': 2,
  '9019': 1,
 
  '3': 
  {
    '0': 'IKEA of Sweden',
    '1': 'TRADFRI bulb GU10 WS 400lm',
    '2': '', '3': '1.2.217',
    '6': 1
  },

  '3311':
  [{
    '5850': 1,
    '5851': 3,
    '5717': 0,
    '5711': 454,
    '9003': 0,
    '5709': 33111,
    '5710': 27211,
    '5706': 'efd275'
  }]
}

>>>
```

**Comment by ferdydek** [comment]
To be fair I'd return it to Ikea and get a replacement unit.

**Comment by mhlund** [comment]
I'm still baffled that it worked at all before factory reset. How could the ios app identify the devices as lights when the coap api returned them as remotes?

**Comment by ferdydek** [comment]
My guess is that ios app uses other method than coap api to communicate? 
Anyway since it worked only after 10 resets and initially buggy... I'd swap it.

**Comment by mhlund** [comment]
My problems could have been a result of me pushing ahead with the pairing to quickly after a reset... When I had almost given up and let the gateway settle for 10 minutes after rebooting the pairing worked!

**Comment by Lakitna** [comment]
So in conclusion: Somehow there was a problem in the pairing process resulting the GU10 bulbs to be recognised as remotes. This resulted in the original error. The solution was to reset the Gateway and re-pair the network. This problem can be identified by looking at the `5750` key, this has a value of `2` when identified as a lamp and `0` when identified as a remote.

While figuring this out @ggravlingen and @lwis made a fix. That fix is now only worth something as a fallback for when this issue appears again.

The IKEA app identifies devices differently than Pytradfri does. IKEA doesn't seem to rely on the value of `5750` at all or on `3311` being present. Instead, the app probably looks at the presence of any lamp values (ex: Lamp state, dimmer value etc.). A curious choice when `5750` exists.

We can automate the detection of this issue, that might be a good addition to the examples.

**Comment by ggravlingen** [comment]
@Lakitna yes, that pretty much sums it up! I looked at your Gist from the other issue and really like it. :+1:  Having seen it, I definitely think it should merged into the master branch. Perhaps as example_debug_info.py to keep in line with the other example scripts?

**Comment by Lakitna** [comment]
Yeah sounds good to me, another solution could be to place it in the folder `script` as most people won't need to use it.

**Comment by ggravlingen** [comment]
@Lakitna I've tried to move the example scripts to a subdirectoriy (I would prefer to not have them in the root dir) but was not able to import pytradfri classes properly. This is probably due to my lack of skills so if you're able to solve it I would appreciate a PR where all examples and this new file is moved to a scripts folder.

**Comment by ggravlingen** [comment]
Can't such a file be added though?

Ping @lwis for architechtural questions.

**Comment by Lakitna** [comment]
Will do! I'll make a PR when I start working on it.

**Comment by andrewerrington** [comment]
I am experiencing something similar.

I have the GU10 full-spectrum bulbs, but they are not recognised as lamps in pytradfri.

The bulbs were bought from Ikea in UK, but I am living in New Zealand. Unfortunately I can't get the Tradfri app because it is region blocked. I have got two GU10 bulbs paired by using the button in the gateway.

Here is the output for one bulb:

devices[3].raw
{'9020': 1520759379, '3': {'2': '', '1': 'TRADFRI bulb GU10
WS 400lm', '3': '1.1.1.1-5.7.2.0', '6': 1, '0': 'IKEA of Sw
den'}, '9054': 1, '9019': 1, '5750': 1, '9001': '', '9002':
1520759371, '9003': 65539}

Here's another one:

devices[2].raw
{'9020': 1520759108, '3': {'2': '', '1': 'TRADFRI bulb GU10
WS 400lm', '3': '1.2.217', '6': 1, '0': 'IKEA of Sweden'},
9054': 0, '9019': 0, '5750': 1, '9001': '', '9002': 1520758
11, '9003': 65538}

Both bulbs are LED1537R6

My gateway is 1.3.14 

I haven't been able to do anything with the gateway yet. Without the app I am unsure if I am pairing properly. I do have a couple of remotes, but I want to get rid of them and do everything via home automation software through the gateway.

**Comment by ferdydek** [comment]
@andrewerrington Seems that the firmware needs update. I dont think you can update it without the app ... so lets think of ways for you to get the app:
Since it wouldn't be smart to advice you to install apk from unknown source, just get an android emulator (ie https://downloadnox.com/pc/) and using browser download the _untrusted_ https://www.apkmonk.com/app/com.ikea.tradfri.lighting/ and install from the downloaded file.
Most likely the discovery will fail and you will need to enter the IP manually, and manually insert serial number and key.

**Comment by andrewerrington** [comment]
Thanks for replying. Which firmware needs updating? The bulb? If item '3' is the firmware version (looks plausible) then I have two different firmware versions.

I was hoping the app wasn't strictly necessary. I'll follow your suggestion to get it. I emailed IKEA and they said yeah, only downloadable in UK, sorry. I hope all the elements of Tradfri are not region-locked. That would be dumb, especially as Tradfri should be coming to IKEA in Australia sometime.

**Comment by ferdydek** [comment]
Yup, 1.2.217 is the current fw version for I think all bulbs.
I don't think the app itself has any limitation, I've tested the method I gave you without giving any location information (or so I think) to the OR or the app.

**Comment by andrewerrington** [comment]
Ok, so one of my bulbs should appear in the 'lights' list in pytradfri. So far only one bulb does. It is an E27 base bulb:

devices[0].raw
{'9001': '', '9019': 0, '3311': [{'5850': 1, '5709': 0, '90
3': 0, '5710': 0, '5851': 20, '5706': '0', '5717': 0, '5711
: 0}], '9002': 1513110441, '9003': 65537, '5750': 2, '3': {
0': 'IKEA of Sweden', '3': '1.1.1.1-5.7.2.0', '6': 1, '2':
', '1': 'TRADFRI bulb E27 WS clear 950lm'}, '9054': 1, '902
': 1513110441}

Unfortunately I can't send any commands to it (change brightness, etc.)

I understand the app is not region-locked, just access to the app in Google Play. I'll get back on to this in a couple of days. Thanks for your help.

**Comment by ggravlingen** [comment]
@andrewerrington at least in theory you should be able to do anything using command line (or pytradfri) that you are able to do using the app. We're using the same methods the app does to do things with the gateway.

There might we ways to force an update. Are you able to run the debug-script in the example folder? If so, what is returned if you pick choice (1), gateway?

**Comment by andrewerrington** [comment]
Ok, so one of my GU10 bulbs should appear in the 'lights' list in pytradfri. So far only one bulb does. It is an E27 base bulb:

devices[0].raw
{'9001': '', '9019': 0, '3311': [{'5850': 1, '5709': 0, '90
3': 0, '5710': 0, '5851': 20, '5706': '0', '5717': 0, '5711
: 0}], '9002': 1513110441, '9003': 65537, '5750': 2, '3': {
0': 'IKEA of Sweden', '3': '1.1.1.1-5.7.2.0', '6': 1, '2':
', '1': 'TRADFRI bulb E27 WS clear 950lm'}, '9054': 1, '902
': 1513110441}

Unfortunately I can't send any commands to it (change brightness, etc.)

I understand the app is not region-locked, just access to the app in Google Play. I'll get back on to this in a couple of days. Thanks for your help.

**Comment by ggravlingen** [comment]
Another thing that worked for me in the past was to put the remote near the bulbs, which led to my updating after a while. Guess the mesh network between the gateway and the bulbs was strengthened this way.

**Comment by andrewerrington** [comment]
Thanks. I'll try that in a couple of days. It's Monday morning in New Zealand now.

**Comment by IVI053** [comment]
I experienced the same as @mhlund. Four of my lamps have the `15009` remote part instead fo `3311`. Will try to reset only the bulbs and pair them another way. Maybe it is related how you pair them.

@mhlund are you using multiple remotes in one group? It look's like that could be part of the problem with my setup as most of the affected devices are in groups with multiple remotes.

As `5750` reports `0` and `2`, I observed that all lamps have `"3":{"6":1}` no matter of `5750`. Remotes (and dimmers) have  `"3":{"6":3}`. Could that be the way IKEA identifies them for the app?

**Comment by andrewerrington** [comment]
Here is the result of running the debug script in the example folder and selecting option 1:
{
    "9023": "3.tradfri.pool.ntp.org",
    "9029": "1.3.14",
    "9054": 0,
    "9055": 0,
    "9059": 1521105117,
    "9060": "2018-03-15T09:11:57.015104Z",
    "9061": 0,
    "9062": 0,
    "9066": 5,
    "9069": 1515442919,
    "9071": 1,
    "9072": 0,
    "9073": 0,
    "9074": 0,
    "9075": 0,
    "9076": 0,
    "9077": 0,
    "9078": 0,
    "9079": 0,
    "9080": 0,
    "9081": "7e133952044000b5",
    "9082": true,
    "9083": "675-63-399",
    "9092": 0,
    "9093": 0,
    "9105": 0,
    "9106": 0
}

**Comment by ggravlingen** [comment]
Looks like you're already on the latest version of the gateway firmware. No automatic updates of the firmware in your bulbs over the past week?

**Comment by andrewerrington** [comment]
No updates. Maybe my gateway is too far from my bulbs (about 14m).

But I don't think that is the issue. I think the issue is this:

>>> devices
[<65537 -  (TRADFRI bulb E27 WS clear 950lm)>, <65536 -  (T
ADFRI remote control)>, <65538 -  (TRADFRI bulb GU10 WS 400
m)>, <65539 -  (TRADFRI bulb GU10 WS 400lm)>]
>>> lights
[<65537 -  (TRADFRI bulb E27 WS clear 950lm)>]
>>>

You can see that I have three bulbs paired with the gateway, but only one of them is categorised as a light.

**Comment by ferdydek** [comment]
If I recall correctly the bulb's fw upgrade from 1.1.xxx to 1.2.214 was not automatic, only from 1.2.214 to 1.2.217 was.

**Comment by mhlund** [comment]
@IVI053 No, I was not using multiple remotes when I had my problems.

**Comment by IVI053** [comment]
I did some more testing and sadly it looks like it had nothing to do with using multiple remotes. @mhlund's situation proves this to.

But I was able to reproduce pairing a bulb with 15009, but no idea why.
It has nothing to do with the device type as far as I can say. I was only able to get a 1.2.214 to report 15009 while testing. Looks like 1.2.217 is only for temperature changing bulbs and 1.3.002 for color ATM. It made no difference how I paired the devices (using remote #1, remote #2 or on gateway directly). Out of 4 pairings I got one time 15009.

@andrewerrington distance can not be a problem in your case. Mine is further away with multiple walls in between and all bulb updates run smoothly. It can take a while for remotes, because of their sleep state. I noticed they update faster if you activate them (turning, button press, motion - depending on type) from time to time.

**Comment by ggravlingen** [comment]
No activity for a while, closing the issue.
